uefi-test-runner/examples/shell_params.rs

```diff
@@ -0,0 +1,60 @@
+// ANCHOR: all
+// ANCHOR: features
+#![no_main]
+#![no_std]
+// ANCHOR_END: features
+
+// ANCHOR: use
+use log::info;
+use uefi::{
+    prelude::*,
+    proto::shell_params::ShellParameters,
+    table::boot::{OpenProtocolAttributes, OpenProtocolParams, SearchType},
+    Identify,
+};
+// ANCHOR_END: use
+
+// ANCHOR: entry
+#[entry]
+fn main(_image_handle: Handle, mut system_table: SystemTable<Boot>) -> Status {
+    // ANCHOR_END: entry
+    // ANCHOR: services
+    uefi_services::init(&mut system_table).unwrap();
+    let boot_services = system_table.boot_services();
+    // ANCHOR_END: services
+
+    // ANCHOR: params
+    let shell_params_h = boot_services
+        .locate_handle_buffer(SearchType::ByProtocol(&ShellParameters::GUID))
+        .unwrap();
+    info!("Found {} ShellParams handles", (*shell_params_h).len());
+    for handle in &*shell_params_h {
+        let params_handle = unsafe {
+            boot_services
+                .open_protocol::<ShellParameters>(
```

Comment:
Can we use `open_protocol_exclusive` here? That will avoid the need for `unsafe`.

Reply:
Yes! And we can open it directly on the image handle, without iterating through all handles.
Otherwise we might open it on the wrong handle. This can happen if the shell is nested.